### PR TITLE
Added task_id to Tests and Test Reports

### PR DESCRIPTION
**Pull request by BethanyG**

### CODE REVIEW

**Comment by BethanyG** [comment]
WIP-ing this until I get the golden tests updated.

**Comment by BethanyG** [comment]
@cmccandless  -  So now all my added tests are passing.  But I am worried.  I had to generate the golden files for my added tests using the `run-in-docker.sh` script, which tells me that we do indeed have different results when using `run.sh` vs `run-in-docker.sh` vs plain pytest.  Exactly the scenario we're trying to avoid.  

We also definitely have a bug when capturing `stdout` in conjunction with a subtest.  It appears that `stdout` is not flushed when using the `subtest` context -- or at least that is not happening when you invoke `run.sh` from the command line.

To see this in action, edit `example_partial_failure_with_subtest.py` to include a `print("Captured stdout from user!!")` statement, and then run `run.sh` for it.  The `results.json` ends up with `output` from all `subtests` written into every `subtest` result.  

I feel like it's not _fatal_ - just really ugly (_and should be fixed_), so I'll log a bug in the morning, and see if I can figure out how to fix it.

**Comment by cmccandless** [comment]
That does sound messy... I love the maintainability of subtests/parameterization, but it's proving to be a real headache with tooling. We might want to discuss the possibility of breaking out the subtests into individual tests.

**Comment by BethanyG** [comment]
Yup.  I don't want to, but if we keep running into issues like this we'll have to.  I am hoping this is a small adjustment. But if not, we will have to either do individual tests, or maybe use PyTest to parameterize, as opposed to `unittest.subtest`.  However, replacing `unittest.subtest` carries a larger refactor of code, since we filter the AST using `unittest` class definitions.

**Comment by BethanyG** [comment]
Huh.  It appears the bug is not just about subtests.  It looks like `stdout` is not being captured when a test fails -- only when it succeeds.  Or in the case of the `subtest` scenario, the parent is collecting the `stdout` from the children.  _sigh_.

**Comment by BethanyG** [comment]
**TL;DR**:  I think this can be reviewed/merged, and then we work on the bugs.

File: runner/__init__.py [hunk 1]
```diff
@@ -81,6 +91,19 @@ def pytest_runtest_logreport(self, report):
...
+            parent_test_name = state.name.split(' ')[0]
+            parent_task_id = self.tests[parent_test_name].task_id
+            state.task_id = parent_task_id
+
+            # Changes status of parent to fail if any of the subtests fail.
+            if state.fail:
+                self.tests[parent_test_name].fail(
+                    message="One or more subtests for this test failed. Details can be found under each variant."
+                )
+                self.tests[parent_test_name].test_code = state.test_code
```
Comment (cmccandless) [inline runner/__init__.py:105]: Does the subtests module not already do this?

Comment (BethanyG) [reply runner/__init__.py:105]: 🤣    No.  The subtests module makes it possible to use `subttest` without PyTest freaking out.  There is actually [an ongoing discussion](https://github.com/pytest-dev/pytest-subtests/issues/11) on the `pytest-subtest` repo on how to handle failures and test counting right now.   And - as the discussion outlines - this is actually behavior inherited from `unittest`.  So for our plugin, I decided to make it "cleaner".  But it is still weird to have a count mis-match when all tests pass vs when some that have sub-tests fail (_this happens with all parameterization, in Unittest/Pytest if I am remembering correctly_).

Comment (BethanyG) [reply runner/__init__.py:105]: ...I'd actually love to do a refactor of our runner that treats parameterization more like a matrix -- so that there is a clean count of which tests are parameterized, then a process that "explodes" the matrix into individual cases, then makes a summary.  But I don't think we want to do that right now.

Comment (cmccandless) [reply runner/__init__.py:105]: No probably right now, but that sounds like a good idea.
